### ISSUE TITLE
pnext: Exception in Administrator when stopping database

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3293]
In latest nightly **pnext** (Version=2.3.588), I'm trying to stop a database with `staradmin stop db` and experience this

```
"ScErrUnexpectedCommandException (SCERR10016): Stopping database harbor. failed due to an unexpected problem.
Version: 2.3.588.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10016.
 System.NullReferenceException: Object reference not set to an instance of an object.
   at Administrator.Server.Managers.ServerManager.ServerModel_Changed(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Managers\ServerManager.cs:line 302
   at Administrator.Server.Model.Server.OnChanged(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Model\Server.cs:line 194
   at Administrator.Server.Model.Database.OnChanged(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 391
   at Administrator.Server.Model.Database.Database_PropertyChanged(Object sender, PropertyChangedEventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 376
   at Administrator.Server.Model.Database.OnPropertyChanged(String fieldName) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 1245
   at Administrator.Server.Model.Database.InvalidateModel() in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 601
   at Administrator.Server.Managers.ServerManager.<>c.<RegisterInternalModelApi>b__8_2(String databaseName, Request request) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\Server\Managers\ServerManager.cs:line 282
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Rest.UriManagedHandlersCodegen.RunUriMatcherAndCallHandler(String methodSpaceUriSpace, String methodSpaceUriSpaceLower, Request req, UInt16 portNumber, Response& resp) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 1114
   at Starcounter.Self.DoSelfNodeCall(UInt16 portNumber, String method, String relativeUri, Dictionary`2 headersDictionary, String body, Byte[] bodyBytes, HandlerOptions handlerOptions, Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Internal\Rest\Self.cs:line 392
   at Starcounter.Self.PUT(String uri, String body, Byte[] bodyBytes, Dictionary`2 headersDictionary, UInt16 port, HandlerOptions ho) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Internal\Rest\Self.cs:line 177
   at Starcounter.Server.PublicModelProvider.UpdateDatabase(Database database) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Server\Starcounter.Server\PublicModelProvider.cs:line 77
   at Starcounter.Server.Commands.Processors.StopDatabaseCommandProcessor.Execute() in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Server\Starcounter.Server\Commands\Processors\StopDatabaseCommandProcessor.cs:line 58
   at Starcounter.Server.Commands.CommandProcessor.ProcessCommand(NotifyCommandStatusChangedCallback notifyStatusChangedCallback) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Server\Starcounter.Server\Commands\CommandProcessor.cs:line 298
HResult=-2147467261
```

followed by this

```
Starcounter.DbException: ScErrUnexpectedCommandException (SCERR10016): Stopping database harbor. failed due to an unexpected problem. ("Object reference not set to an instance of an object.").
Version: 2.3.588.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10016.
   at Starcounter.Administrator.API.Handlers.EngineHandler.ToErrorResponse(CommandInfo commandInfo) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\API\Handlers\EngineHandler.DELETE.cs:line 115
   at Starcounter.Administrator.API.Handlers.EngineHandler.OnDELETE(String name, Request request) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Administrator\API\Handlers\EngineHandler.DELETE.cs:line 53
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-38\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 557
ErrorCode=10016
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR10016
```

Unfortunately, because of #3261, I can't debug it at this point. Its not deterministic; it happens every time.

**Comment by chrhol** [comment]
That's strange, some of the tests that are currently running for nightly build uses `staradmin stop db` command to stop the database before deleting it. Look at for example `ClassSchemaChange_test.bat`, and this test is succesful for build 2.3.588

Is it some special database  (or maybe default) that fails?

**Comment by per-samuelsson** [comment]
> Is it some special database (or maybe default) that fails?

A custom one:

> Stopping database harbor

Basically the only special thing about it is that it's configured to be a replica, in a cluster of 3 databases, but I doubt that means anything to the Administrator API's. :wink: 

After I restarted all databases, and now tried again, I don't experience it. Let's keep an eye on it. I'll update here when/if I experience it again.

**Comment by careri** [comment]
I get the same error:
```
ScErrUnexpectedCommandException (SCERR10016): Starting HAD in had failed due to an unexpected problem.
Version: 2.1.1116.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10016.
 System.NullReferenceException: Object reference not set to an instance of an object.
```

**Comment by chrhol** [comment]
@careri: The problem should be fixed by #3298. The latest available nightly (2.1.1123) should include the fix.

@per-samuelsson: I merged the fix to pnext. I let you decide when and if we can close this issue.

**Comment by per-samuelsson** [comment]
> @per-samuelsson: I merged the fix to pnext. I let you decide when and if we can close this issue.

It's still there: see my recent https://github.com/Starcounter/Starcounter/issues/3298#issuecomment-175029947.

**Comment by dan31** [comment]
Do we continue to meet this? I was not able to reproduce.

**Comment by per-samuelsson** [comment]
I just closed #3298, and I'm closing this on the same premises.
